enhancements/network/ovs-hardware-offload.md

```diff
@@ -0,0 +1,242 @@
+---
+title: ovs-hardware-offload
+authors:
+  - "@zshi"
+reviewers:
+  - "@dcbw"
+  - "@squeed"
+  - "@danwinship"
+  - "@pliurh"
+  - "@dougbtv"
+  - "@s1061123"
+approvers:
+  - "@knobunc"
+  - "@fepan"
+creation-date: 2020-04-27
+last-updated: 2020-04-27
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+
+---
+
+# OVS-Hardware-Offload
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [X] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+- When will RHCOS uses RHEL-8.3 as base image?
+- Any known issues with connction track feature using OVS Hardware Offload?
+- Any known issues with IPv6 traffic using OVS Hardware Offload?
+
+## Summary
+
+This enhancement is to introduce OVS hardware offload feature in OpenShift
+for high-performance datapath with OVN-Kubernetes. OVS hardware offload
+enables offloading of OVS datapath rules to a compatible network card using
+linux Traffic Control (TC) tool `tc`. It relies on SR-IOV technology to control
+the SR-IOV Virtual Functions (VF) via VF representor port that is plugged
+into OpenvSwitch bridge. It improves the utilization of available bandwidth
+on underlying NICs by moving CPU intensive network packet processing from
+OVS software to hardware.
+
+## Motivation
+
+Enable high-performance datapath on primary pod interface with production-ready
+OVN-Kubernetes.
+
+### Goals
+
+Enable OpenFlow hardware offload for OVN datapath
+
+### Non-Goals
+
+Replace OpenFlow software based solution
+
+## Proposal
+
+OVS Hardware Offload accelerates the OVS software flows with a SR-IOV NIC
+that works in `switchdev` mode. It requires configurations in both OpenvSwitch
+daemon and SR-IOV hardware NIC.
+
+- OpenvSwitch daemon is managed by Cluster Network Operator (CNO) and runs in
+`ovs-node` DaemonSet. A boolean configurable field will be added in CNO
```

Comment:
Is this boolean node specific? AFAIK, it is not mandatory to have all the nodes with offload and ovs running on the specific node should have the offload config enabled.

Reply:
It will be a cluster-wide configuration for ovn-kubernetes.
I think it's worth discussion whether we want per-node or per-role control of offload enablement.
One possible way of doing per-role control is to add a pool selector in `spec.defaultNetwork.ovnKubernetesConfig` to specify which MachineConfigPool shall be enabled with hw-offload.

Reply:
Ovs hw offload will be supported on primary pod interface with SR-IOV device.  Having second interface in one OVS instance implies that `ovn-k8s-cni-overlay` be used for additional pod interface, which is not considered in this enhancement.
Traffic will only be offloaded when SR-IOV VF (in switchdev mode) is attached to pod primary network. so ovs hw offload can be request in pod basis by using customized default net-attach-def.

Reply:
@moshe010 For a non-switchdev pod that uses veth pair as primary interface, how would OVS decides not to offload the traffic?  for example, does it check whether the hw-tc-offload feature is enabled or not on the attached device (veth or VF representor)?

Reply:
Rethinking on this, it might make sense to enable ovs hw offload for a specific MachineConfigPool instead of the whole cluster, to isolate the potential impact of "the whole cluster is down due to hw-offload failure" and "not all cluster nodes have hw offload capable NICs".

@pliurh @krsacme @moshe010 @danwinship wdyt?

Reply:
Yes, it should be configurable on a node group rather than the whole cluster, to allow mixed capability on the cluster.

Reply:
> @moshe010 For a non-switchdev pod that uses veth pair as primary interface, how would OVS decides not to offload the traffic? for example, does it check whether the hw-tc-offload feature is enabled or not on the attached device (veth or VF representor)?

When OVS is configured with hardware offload, it uses TC as its default datapath. if the DP rule cannot be handled in TC then OVS kernel DP is used.

in TC, if the underlying device is capable of hardware offload it will try to offload the rule to the device (unless otherwise specified in tc-policy , i.e skip-sw / skip-hw) for VETH device, as its not capable of hardware offloads, the flow will not be offloaded (but would still be handled by TC)

+1 on a per-node group configurations for it.

Reply:
> When OVS is configured with hardware offload, it uses TC as its default datapath. if the DP rule cannot be handled in TC then OVS kernel DP is used.
> 
> in TC, if the underlying device is capable of hardware offload it will try to offload the rule to the device (unless otherwise specified in tc-policy , i.e skip-sw / skip-hw) for VETH device, as its not capable of hardware offloads, the flow will not be offloaded (but would still be handled by TC)

I was thinking to have the default tc-policy configured to skip-sw so that OVS will do either hw offload or kernel datapath until the TC datapath is properly validated and exposing tc-policy as a confiugrable option in case people want to try tc software datapath.
> 
> +1 on a per-node group configurations for it. i would not move the entire cluster to use OVS with TC Datapath until it was properly validated.

To identify a group of nodes, we have two options:
one is to use existing MachineConfigPool, the other is to define customized labels for a node group.
The first group may not contains the exact hw-offload capable nodes (some nodes may have capable NIC, others may not as the pool is not grouped according to hw-offload capability). The second would allow us to accurately define a hw-offload capable node pool. I think the second is what we should do.